go/client/cmd_chat_upload.go

```diff
@@ -2,61 +2,171 @@ package client
 
 import (
 	"context"
-	"errors"
+	"fmt"
+	"os"
 
 	"github.com/keybase/cli"
 	"github.com/keybase/client/go/libcmdline"
 	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/chat1"
+	"github.com/keybase/go-framed-msgpack-rpc/rpc"
 )
 
 type CmdChatUpload struct {
 	libkb.Contextified
-	tlf      string
-	filename string
-	public   bool
+	filename         string
+	useStdin         bool
+	size             int
+	source           Source
+	resolvingRequest chatConversationResolvingRequest
 }
 
 func newCmdChatUpload(cl *libcmdline.CommandLine, g *libkb.GlobalContext) cli.Command {
 	return cli.Command{
 		Name:         "upload",
 		Usage:        "Upload an attachment to a conversation",
-		ArgumentHelp: "<conversation> <filename>",
+		ArgumentHelp: "[conversation [filename]]",
 		Action: func(c *cli.Context) {
 			cmd := &CmdChatUpload{Contextified: libkb.NewContextified(g)}
 			cl.ChooseCommand(cmd, "upload", c)
 		},
-		Flags: []cli.Flag{
-			cli.BoolFlag{
-				Name:  "public",
-				Usage: "Send to public conversation (default private)",
-			},
-		},
+		Flags: append(getConversationResolverFlags(),
+			mustGetChatFlags("stdin", "filename")...,
+		),
 	}
 }
 
-func (c *CmdChatUpload) ParseArgv(ctx *cli.Context) error {
-	if len(ctx.Args()) != 2 {
-		return errors.New("usage: keybase chat upload <conversation> <filename>")
+func (c *CmdChatUpload) useFileSource(filePath string) (err error) {
+	c.source = NewFileSource(filePath)
+	info, err := os.Stat(filePath)
+	if err != nil {
+		return err
+	}
+	if info.IsDir() {
+		return fmt.Errorf("%s is a directory", filePath)
+	}
+	if len(c.filename) == 0 {
+		c.filename = info.Name()
+		c.size = int(info.Size())
+	}
+	return nil
+}
+
+func (c *CmdChatUpload) ParseArgv(ctx *cli.Context) (err error) {
+	c.useStdin = ctx.Bool("stdin")
+	c.filename = ctx.String("filename")
+
+	var tlfName string
+	// Get the TLF name from the first position arg
+	if len(ctx.Args()) >= 1 {
+		tlfName = ctx.Args().Get(0)
+	}
+	if c.resolvingRequest, err = parseConversationResolvingRequest(ctx, tlfName); err != nil {
+		return err
+	}
+
+	if c.useStdin {
+		if len(ctx.Args()) != 1 {
+			return fmt.Errorf("need exactly 1 argument to upload from stdin")
+		}
+		if len(c.filename) == 0 {
+			return fmt.Errorf("need --filename to upload from stdin")
+		}
+		c.source = &StdinSource{}
+	} else {
+		switch len(ctx.Args()) {
+		case 0, 1:
+			c.source = nil
+		case 2:
+			if err = c.useFileSource(ctx.Args().Get(1)); err != nil {
+				return err
+			}
+		default:
+			cli.ShowCommandHelp(ctx, "upload")
+			return fmt.Errorf("chat upload takes 1 or 2 args")
+		}
 	}
-	c.tlf = ctx.Args()[0]
-	c.filename = ctx.Args()[1]
-	c.public = ctx.Bool("public")
 
 	return nil
 }
 
 func (c *CmdChatUpload) Run() error {
-	opts := attachOptionsV1{
-		Channel: ChatChannel{
-			Name:   c.tlf,
-			Public: c.public,
-		},
+	chatClient, err := GetChatLocalClient(c.G())
```

Comment:
Why do you need the resolver?  Just prompt for whatever is necessary and fill in attachOptionsV1 with the prompt results instead of flags:

```	opts := attachOptionsV1{
		Channel: ChatChannel{
			Name:   c.tlf,
			Public: c.public,
		},
		Filename: c.filename,
	}
	h := newChatServiceHandler(c.G())
	reply := h.AttachV1(context.Background(), opts)
	if reply.Error != nil {
		return errors.New(reply.Error.Message)
	}
	return nil
```